AudioKit/Common/Nodes/Playback/Player/AKPlayer.swift

```diff
@@ -0,0 +1,724 @@
+
+        // for decay to x% amplitude (-dB) over the given decay time
+        let fadeOutPower: Double = exp(-log(25) * sampleTime / fade.outTime)
+
+        // where in the buffer to end the fade in
+        let fadeInSamples = Int(audioFile.sampleRate * fade.inTime)
+        // where in the buffer to start the fade out
+        let fadeOutSamples = Int(Double(length) - (audioFile.sampleRate * fade.outTime))
+
+        // i is the index in the buffer
+        for i in 0 ..< Int(length) {
+            // n is the channel
+            for n in 0 ..< Int(buffer.format.channelCount) {
+
+                if i <= fadeInSamples && fade.inTime > 0 {
+                    gain *= fadeInPower
+                } else if i >= fadeOutSamples && fade.outTime > 0 {
+                    if i == fadeOutSamples {
+                        gain = fade.outStartGain
+                    }
+                    gain *= fadeOutPower
+                } else {
+                    gain = 1.0
+                }
+
+                //sanity check
+                if gain > 1 {
+                    gain = 1
+                }
+
+                let sample = buffer.floatChannelData![n][i] * Float(gain)
+                fadedBuffer?.floatChannelData?[n][i] = sample
+            }
+        }
+
+        // set the buffer now to be the faded one
+        self.buffer = fadedBuffer
+        // update this
+        self.buffer?.frameLength = length
+    }
+
+    // Read the buffer in backwards
+    fileprivate func reverseBuffer() {
+        if buffer == nil {
+            //updateBuffer()
+        }
+        guard isBuffered, let buffer = self.buffer else { return }
+
+        let reversedBuffer = AVAudioPCMBuffer(
+            pcmFormat: buffer.format,
+            frameCapacity: buffer.frameCapacity
+        )
+
+        var j: Int = 0
+        let length = buffer.frameLength
+        //AKLog("reverse() preparing \(length) frames")
+
+        // i represents the normal buffer read in reverse
+        for i in (0 ..< Int(length)).reversed() {
+            // n is the channel
+            for n in 0 ..< Int(buffer.format.channelCount) {
+                // we write the reverseBuffer via the j index
+                reversedBuffer?.floatChannelData?[n][j] = buffer.floatChannelData?[n][i] ?? 0.0
+            }
+            j += 1
+        }
+
+        reversedBuffer?.frameLength = length
+
+        // set the buffer now to be the reverse one
+        self.buffer = reversedBuffer
+
+        AKLog("Buffer reversed: \(isReversed)")
+    }
+
+    // Disconnect the node
+    override open func disconnect() {
+        stop()
+        audioFile = nil
+        buffer = nil
+        AudioKit.detach(nodes: [mixer, playerNode])
+    }
+
+    deinit {
+        AKLog("* deinit AKPlayer")
+    }
+
+    // MARK: - Static Methods
+
+    /// convert an AVAudioTime object to seconds with a hostTime reference
+    open class func audioTimeToSeconds(hostTime: UInt64, audioTime: AVAudioTime) -> Double {
```

Comment:
This should return an optional Double and return nil if !isHostTimeValid.
```if !audioTime.isHostTimeValid { return nil }```

Reply:
Also, it should be moved from this class as it's unrelated to the player.  Add it as an extension to AVAudioTime in AudioKit/Common/Internals/Utilities/AVAudioTimeShim.swift.  Or make it private.  Another option, which you may prefer, is to create an AKUtility class and add it as a static function there.